src/phylo/nodes.jl

```diff
@@ -0,0 +1,655 @@
+  end
+end
+
+@Docile.doc """
+Prune a PhyNode from its parent and graft it to another parent.
+
+**Parameters:**
+
+* `prune`:    The PhyNode to remove from its parent.
+
+* `graftto`:  The PhyNode to become the new parent of `prune`.
+""" ->
+function pruneregraft!(prune::PhyNode, graftto::PhyNode)
+  x = prune!(prune)
+  graft!(graftto, x)
+end
+
+@Docile.doc """
+Prune a PhyNode from its parent and graft it to another parent, setting the branch length.
+
+**Parameters:**
+
+* `prune`:        The PhyNode to remove from its parent.
+
+* `graftto`:      The PhyNode to become the new parent of `prune`.
+
+* `branchlength`: The branch length.
+""" ->
+function pruneregraft!(prune::PhyNode, graftto::PhyNode, branchlength::Float64)
+  x = prune!(prune)
+  graft!(graftto, x, branchlength)
+end
+
+@Docile.doc """
+Delete a node, destroying the relationships between it and its parent, and it and its children. The children of the node become the children of the node's former parent.
+
+Returns the deleted node.
+
+**Parameter:**
+
+* `x`: The PhyNode to delete.
+""" ->
+function delete!(x::PhyNode)
+  deleted = prune!(x)
+  graft!(parent(deleted), children(deleted))
+  return deleted
+end
+
+@Docile.doc """
+Detach a subtree at a given node.
+
+Returns a new Phylogeny with the detached node as root.
+
+**Parameters:**
+
+* `x`:          The PhyNode to detach.
+
+* `name`:       The name of the new Phylogeny.
+
+* `rooted`:     Whether the detached subtree is rooted.
+
+* `rerootable`: Whether the detached subtree is rerootable.
+""" ->
+function detach!(x::PhyNode, name::String = "", rooted::Bool = true, rerootable::Bool = true)
+  detached = prune!(x)
+  return Phylogeny(name, detached, rooted, rerootable)
+end
+
+@Docile.doc """
+Test whether two PhyNodes are equal. Specifically, test whether all three of `branchlength`, `name` and `extensions` are equal.
+
+**Parameters:**
+
+* `x`: The left PhyNode to compare.
+
+* `y`: The right PhyNode to compare.
+""" ->
+function isequal(x::PhyNode, y::PhyNode)
+  bl = x.branchlength == y.branchlength
+  n = x.name == y.name
+  return all([bl, n])
+end
+
+@Docile.doc """
+Find the distance of a node from its parent. This is different from branch length, because it handles the situation where branch length is unknown. It is only used when the distances between nodes are calculated.
+
+The method is necessary because unknown branch lengths are represented as -1.0.
+If all branch lengths are unknown, the tree is a cladogram, and it is still useful to be able to compare relative distances. If individual branch lengths are unknown, they should not affect the calculation of path distances. To satisfy both of these cases, we use machine epsilon as the minimal distance.
+
+**Parameters:**
+
+* `x`: A PhyNode.
+
+""" ->
+function distanceof(x::PhyNode)
+  bl = branchlength(x)
+  return bl == -1.0 ? eps() : bl
```

Comment:
I can see how returning `eps()` on unknown branch length could be useful, but it seems like it could also be very surprising. You could have two versions, one that errors by default, and another that replaces unknown values with some specified value, like:
```julia
function distanceof(x::PhyNode)
    b1 = branchlength(x)
    if b1 == -1.0
        error("Branch length is unknown.")
    end
    return b1
end

function distanceof(x::PhyNode, replace_unknown::Float64)
    b1 = branchlength(x)
    if b1 == -1.0
        return replace_unknown
    end
    return b1
end
```